### PR TITLE
- make the split collective shared file pointer operations work

### PR DESCRIPTION
**Pull request by edgargabriel**
- minor code restructering in io/ompio required for that.

### CODE REVIEW

**Comment by edgargabriel** [comment]
@hppritcha this is now the cleaned up version of the split collective operations, not containing any other fixes. If you have a chance to review it, please go ahead. If you prefer to review the pr to v2.x, let me know and I can merge this commit and file the pr.

File: ompi/mca/sharedfp/individual/sharedfp_individual_iwrite.c [hunk 1]
```diff
@@ -90,14 +90,132 @@ int mca_sharedfp_individual_write_ordered_begin(mca_io_ompio_file_t *fh,
...
+                                                        fh->f_info,
+                                                        fh);
+        if ( OMPI_SUCCESS != ret ) {
+            opal_output(0,"sharedfp_individual_write_ordered_begin - error opening the shared file pointer\n");
+            return ret;
+        }
+    }
+
+    if ( true == fh->f_split_coll_in_use ) {
+        printf("Only one split collective I/O operation allowed per file handle at any given point in time!\n");
```
Comment (hppritcha) [inline ompi/mca/sharedfp/individual/sharedfp_individual_iwrite.c]: should this be using opal_output instead?

File: ompi/mca/sharedfp/lockedfile/sharedfp_lockedfile_iread.c [hunk 2]
```diff
@@ -87,8 +87,121 @@ int mca_sharedfp_lockedfile_read_ordered_begin(mca_io_ompio_file_t *fh,
...
+                                                        fh);
+        if ( OMPI_SUCCESS != ret ) {
+            opal_output(0,"sharedfp_lockedfile_read_ordered_begin - error opening the shared file pointer\n");
+            return ret;
+        }
+    }
+
+
+    if ( true == fh->f_split_coll_in_use ) {
+        printf("Only one split collective I/O operation allowed per file handle at any given point in time!\n");
```
Comment (hppritcha) [inline ompi/mca/sharedfp/lockedfile/sharedfp_lockedfile_iread.c]: opal_output here too?

File: ompi/mca/sharedfp/lockedfile/sharedfp_lockedfile_iwrite.c [hunk 3]
```diff
@@ -85,8 +85,121 @@ int mca_sharedfp_lockedfile_write_ordered_begin(mca_io_ompio_file_t *fh,
...
+                                                        fh);
+        if ( OMPI_SUCCESS != ret ) {
+            opal_output(0,"sharedfp_lockedfile_write_ordered_begin - error opening the shared file pointer\n");
+            return ret;
+        }
+    }
+
+
+    if ( true == fh->f_split_coll_in_use ) {
+        printf("Only one split collective I/O operation allowed per file handle at any given point in time!\n");
```
Comment (hppritcha) [inline ompi/mca/sharedfp/lockedfile/sharedfp_lockedfile_iwrite.c]: opal_output here too?

File: ompi/mca/sharedfp/sm/sharedfp_sm_iread.c [hunk 4]
```diff
@@ -84,8 +84,123 @@ int mca_sharedfp_sm_read_ordered_begin(mca_io_ompio_file_t *fh,
...
+                                                        fh->f_amode,
+                                                        fh->f_info,
+                                                        fh);
+        if (  OMPI_SUCCESS != ret ) {
+            opal_output(0,"sharedfp_sm_read_ordered_begin - error opening the shared file pointer\n");
+            return ret;
+        }
+    }
+    if ( true == fh->f_split_coll_in_use ) {
+        printf("Only one split collective I/O operation allowed per file handle at any given point in time!\n");
```
Comment (hppritcha) [inline ompi/mca/sharedfp/sm/sharedfp_sm_iread.c]: opal_output?

File: ompi/mca/sharedfp/sm/sharedfp_sm_iwrite.c [hunk 5]
```diff
@@ -85,8 +85,123 @@ int mca_sharedfp_sm_write_ordered_begin(mca_io_ompio_file_t *fh,
...
+                                                        fh->f_amode,
+                                                        fh->f_info,
+                                                        fh);
+        if (  OMPI_SUCCESS != ret ) {
+            opal_output(0,"sharedfp_sm_write_ordered_begin - error opening the shared file pointer\n");
+            return ret;
+        }
+    }
+    if ( true == fh->f_split_coll_in_use ) {
+        printf("Only one split collective I/O operation allowed per file handle at any given point in time!\n");
```
Comment (hppritcha) [inline ompi/mca/sharedfp/sm/sharedfp_sm_iwrite.c]: opal_output?

**Comment by hppritcha** [comment]
I'm getting segfaults with filetest.  Please don't merge yet.

**Comment by edgargabriel** [comment]
ok, will also incorporate your comments about using opal_output instead of printf tomorrow morning. You are testing on the cray? Can try on the stuttgart machine as well

**Comment by hppritcha** [comment]
i don't think this problem is specific to cray, but yes I"m using the NERSC systems hopper and edison.

**Comment by hppritcha** [comment]
If I don't use lustre, then filetest passes.  By the way, what's going on with ```mca_fs_lustre_file_get_size ```?

**Comment by hppritcha** [comment]
I'm okay with this PR, but it appears that for Lustre things are pretty broken.  Filetest fails at the very 
first ```ompio_io_ompio_file_close``` owing to some mishandling of the f_converter at line 369.  That's not the whole problem though because if I comment out that line, things still end up segfaulting later in the tests.  The signature is what one would get with heap corruption.

Edison is using Lustre 2.5.0 (at least for the clients).
Hopper is using Lustre 2.4.1

I observe the same behavior on both systems' lustre file systems.

I'm pretty sure things were working for lustre several weeks ago when I tested one of the previous ompi i/o PRs.   I'll try to do some bisecting to narrow down the problem when I have time.

**Comment by edgargabriel** [comment]
hm, ok, thanks! I will merge the branch and check on lustre as well. One simple test that you could run is to see whether things execute correctly if you exclude the lustre fs component, it should still work correctly, e.g.

mpirun --mca fs ^lustre -np 6 ./filetest

Also note, that I did modifications to the filetest testsuite in the last couple of days, maybe I introduced a bug there inadvertently.

**Comment by edgargabriel** [comment]
well, I will first fix the printf to opal_output stuff tomorrow morning, than I will merge.

**Comment by edgargabriel** [comment]
Will have to check mca_fs_lustre_file_get_size, did not look into that in a looong time

**Comment by edgargabriel** [comment]
Can confirm that I am able to reproduce the problem that you see on the Stuttgart Cray/Lustre, and am debugging it. I know at least part of the problem, it should be however unrelated to this pr.

**Comment by edgargabriel** [comment]
the problem definitely comes from the lustre fs component. If I exclude it (on the lustre system) and use the regular ufs component instead, everything works like a charm. I will merge this pr, and debug the lustre fs component to see what is going on. Thanks @hppritcha !

**Pull request merged**